### ISSUE TITLE
Instability with small feature values.

### ISSUE THREAD

**Issue by TorbenVogt** [dmlc/xgboost#2939]
## Environment info
Operating System:
Windows 7 enterprise

Package used (python/R/jvm/C++):
Python

`xgboost` version used:
0.6

The python version and distribution:
Python 2.7.13, Anaconda custom 64-bit <default, May 11 2017>

Python packages:
('NumPy', '1.12.1')
('SciPy', '0.19.0')
('Scikit-Learn', '0.19.0')

## Steps to reproduce
I have troubles when feature values become small but within the floating-point boundaries. XGBoost does not seem to learn.

The issue can be reproduced with the following code. I used the breast_cancer data and multiply all feature values with e^-49 meaning all values are smaller. A test split is then created from the data and the same for data scaled by MinMaxScalor. 
Then XGBoost is used for training and the feature importance and accuracy is printed.

```python
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn import datasets
from xgboost import XGBClassifier

data = datasets.load_breast_cancer()
X = data.data[:, :30]  #
X *= 10**(-49)
print X

X_scaled = MinMaxScaler().fit_transform(X)
y = data.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=7)
X_train_s, X_test_s, y_train_s, y_test_s = train_test_split(X_scaled, y, test_size=0.2, random_state=7)

eval_set = [(X_train_s, y_train_s), (X_test_s, y_test_s)]
model = XGBClassifier(random_state=42)
model.fit(X_train_s, y_train_s, eval_metric=["error", "logloss"], eval_set=eval_set, verbose=False)
print model.feature_importances_
y_pred = model.predict(X_test_s)
accuracy = accuracy_score(y_test_s, y_pred)
print("Accuracy: %.2f%%" % (accuracy * 100.0))

eval_set = [(X_train, y_train), (X_test, y_test)]
model = XGBClassifier(random_state=42)
model.fit(X_train, y_train, eval_metric=["error", "logloss"], eval_set=eval_set, verbose=False)
print model.feature_importances_
y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy: %.2f%%" % (accuracy * 100.0))
```
When using the scaled data for tranning XGBoost does not learn, it gives every feature importance a value of 'nan' and yields an error when trying to calculate the feature importance. Using the MinMaxScalor it yields an accuracy of 97.37% Which is similar to accuracy when simply training on the data.

The error messages and printed output is:

Feature importances - scaled data
[ 0.          0.07391305  0.          0.          0.01956522  0.02173913
  0.00652174  0.06304348  0.          0.00652174  0.02391304  0.00652174
  0.00434783  0.08260869  0.01521739  0.02826087  0.00869565  0.00652174
  0.03043478  0.0326087   0.04130435  0.1326087   0.03913043  0.11086956
  0.04782609  0.01304348  0.05652174  0.09347826  0.02826087  0.00652174]
Accuracy: 97.37%

...\Continuum\Anaconda2\lib\site-packages\xgboost-0.6-py2.7.egg\xgboost\sklearn.py:385: RuntimeWarning: invalid value encountered in divide return all_features / all_features.sum()
Feature importances
[ nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan
  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan  nan]
Accuracy: 64.91%

## What have you tried?
Tried to look into the source code to find an explanation for this but without success.

**Comment by sejlermonster** [comment]
This looks interesting and reproducible. Is this caused by some optimization threshold?

**Comment by khotilov** [comment]
There is a constant `rt_eps = 1e-6f` which is used to compare loss values and give some margins when looking for branch split points
https://github.com/dmlc/xgboost/blob/d9d5293cdbbbf67dc8ff9d4a3f171d0990fdd1ee/include/xgboost/base.h#L203
So, yes, it's not possible to use arbitrarily small values of features.

**Comment by sejlermonster** [comment]
Is this necessary for optimization reasons? Doesn't such a low precision affect the accuracy of the trained model even on more "normal" data sets?

**Comment by khotilov** [comment]
I'm sorry, I keep misremembering what rt_eps does. Case in point: https://github.com/dmlc/xgboost/pull/1940#issuecomment-272343663 , where I was actually making sure that it wasn't used in a way that breaks small features.

The example from @TorbenVogt fails because xgboost's features are stored in 32 bit float precision, while the factor of 10**(-49) makes X to fall under the float resolution. You can see it from ` X.astype('float32').max()`

**Comment by ardabeyazoglu** [comment]
I'm having the same issue with XGBRegressor. Will there be any solution about this ?
